opm/material/densead/DynamicEvaluation.hpp

```diff
@@ -0,0 +1,515 @@
+  (at your option) any later version.
+
+  OPM is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with OPM.  If not, see <http://www.gnu.org/licenses/>.
+
+  Consult the COPYING file in the top-level source directory of this
+  module for the precise wording of the license and the list of
+  copyright holders.
+*/
+/*!
+ * \file
+ *
+ * \brief This file file provides a dense-AD Evaluation class where the 
+ *        number of derivatives is specified at runtime.
+ *
+ * \attention THIS FILE GETS AUTOMATICALLY GENERATED BY THE "genEvalSpecializations.py"
+ *            SCRIPT. DO NOT EDIT IT MANUALLY!
+ */
+#ifndef OPM_DENSEAD_EVALUATION_DYNAMIC_HPP
+#define OPM_DENSEAD_EVALUATION_DYNAMIC_HPP
+
+#include "Evaluation.hpp"
+#include "Math.hpp"
+
+#include <opm/material/common/Valgrind.hpp>
+
+#include <vector>
+#include <cmath>
+#include <cassert>
+#include <cstring>
+#include <iostream>
+#include <algorithm>
+
+namespace Opm {
+namespace DenseAd {
+
+/*!
+ * \brief Represents a function evaluation and its derivatives w.r.t. a 
+ *        run-time specified set of variables.
+ */
+template <class ValueT>
+class Evaluation<ValueT, -1>
+{
+public:
+    //! the template argument which specifies the number of
+    //! derivatives (-1 means runtime determined)
+    static const int numVars = -1;
+
+    //! field type
+    typedef ValueT ValueType;
+
+    //! number of derivatives
+    int size() const
+    { return data_.size() - 1; }
+
+protected:
+    //! length of internal data vector
+    int length_() const
+    { return data_.size(); }
+
+
+    //! position index for value
+    constexpr int valuepos_() const
+    { return 0; }
+    //! start index for derivatives
+    constexpr int dstart_() const
+    { return 1; }
+    //! end+1 index for derivatives
+    int dend_() const
+    { return length_(); }
+
+    //! instruct valgrind to check that the value and all derivatives of the
+    //! Evaluation object are well-defined.
+    void checkDefined_() const
+    {
+#ifndef NDEBUG
+       for (const auto& v: data_)
+           Valgrind::CheckDefined(v);
+#endif
+    }
+
+public:
+    //! default constructor
+    Evaluation() : data_()
+    {}
+
+    //! copy other function evaluation
+    Evaluation(const Evaluation& other) = default;
+
+    // create a "blank" dynamic evaluation
+    explicit Evaluation(int numDerivatives)
+        : data_(1 + numDerivatives)
+    {}
+
+    // create a dynamic evaluation which represents a constant function
+    //
+    // i.e., f(x) = c. this implies an evaluation with the given value and all
+    // derivatives being zero.
+    template <class RhsValueType>
+    Evaluation(int numDerivatives, const RhsValueType& c)
```

Comment:
It is just some personal preference. I would prefer Evaluation(const RhsValueType& c, const int numDerivatives). It is kind of more natural to me to value first, then number of derivatives, since it is also the order in the `data_`.  But  I do not insist this. It is just pure personal preference.

Reply:
this leads to the problem that creating "primitive variables" for static Evaluations has the same signature as dynamically-sized constants. IMO this would be terribly confusing. Also, the STL classes like `std::vector` use (size, value) as their constructor arguments...